### ISSUE TITLE
cleaning/modularizing the canopy's vertical leaf nitrogen effects

### ISSUE THREAD

**Issue by rgknox** [NGEET/fates#622]
Several of us were trying to understand exactly how the vertical scaling of leaf nitrogen concentrations were impacting leaf maintenance respiration.

After some back and forth, we decided that the current scheme isn't necessarily wrong, but it was lacking some features or concepts that could be improved upon.  Here is roughly the part of the code that I'm talking about:

https://github.com/NGEET/fates/blob/master/biogeophys/FatesPlantRespPhotosynthMod.F90#L433-L474

Here are some features or improvments that were brought up:

1) More modularity, and the ability to test hypotheses (ie some argue that nitrogen concentration per unit mass does not vary with depth)
2) More clarity in units.
3) Enforcing mass conservation?  We scale lmr by nscaler, but its not clear that if this nscalar is supposed to be mass conservative, ie it literaly scales nutrient conentrations, or if it is supposed to reflect also the nitrogen effect on LMR over depth
4) We calculate a lmr25top, wich immediately uses slatop.  The calculations never end up using slamax or the actual profile of sla. I suppose that nscalar currently is supposed to be a catch all here, but again this is not conservative, and also does not clearly break apart  how lmr is impacted by 1) vertcial changes in leaf nitrogen content per unit mass of carbon, and 2) vertical changes in leaf area per unit mass of carbon (sla)

@serbinsh @glemieux @walkeranthonyp @ckoven @climate-dude

**Comment by glemieux** [comment]
@rgknox for the modularity, I was seeing code preceding the above part of the code overlap in the `trim_canopy` subroutine, specifically in the calculation of the cumulative lai:
https://github.com/NGEET/fates/blob/ab1647a66ade145104433db8a624c9b53a509db3/biogeochem/EDPhysiologyMod.F90#L447-L465

I'm wondering if this could/should also be wrapped up into a separate `nscalar` calling function itself separate from the lmr25top calculation?

**Comment by rgknox** [comment]
I agree @glemieux , we don't want code repetition.  If any algorithm is used more than once, it shouldn't be re-written, we should have a function or subroutine for it.

**Comment by glemieux** [comment]
I'm going to start a branch to modularize some of these things @rgknox, unless you've got something in the works already.

**Comment by jenniferholm** [comment]
@rgknox - This seems like the same place in code where I might be running into troubles. Do you think some of this vertical scaling for lmr, is why I'm later getting NaNs for MR, AR, leaf_npp?

I'm wondering if because I changed values for prt_nitr_stoich_p1 and prt_nitr_stoich_p2, and now they aren't identical anymore. If p1 and p2 differ, then that messes with the cn_scalar, right? Do the cn_scalar and nscalar interact?  I could be confused and not really understand where the cnp_flex_allom_hyp case comes into play. 

Ryan and I also noticed a typo here: 
https://github.com/NGEET/fates/blob/3b6ac22180b0cc873005a141d7967c4fa9fd6b45/biogeophys/FatesPlantRespPhotosynthMod.F90#L459

and it should be (leaf_organ, nitrogen_element)

**Comment by glemieux** [comment]
Here's the branch that I've started to refactor and modularize this code: https://github.com/glemieux/fates/commits/nscalar-modularization

I was unsure weather or not to include `hlm_use_planthydro` as in input to the new `cumulative_lai` since it appears we compute the cumulative lai differently depending on that parameter switch: https://github.com/glemieux/fates/blob/9a65211558c2da8a0a28d2725a47908729c58af9/biogeophys/FatesPlantRespPhotosynthMod.F90#L398-L418

For now, I just decided to keep the alternate cumulative lai calculation in the above section just to be explicit.  That said, it did make me wonder if that alternate method should be used for the `trim_canopy` function as well:
https://github.com/glemieux/fates/blob/d673ba21c62f3617f2d8c8df42ff7199ad9f0ab0/biogeochem/EDPhysiologyMod.F90#L435-L436
